src/core/libraries/usbd/usb_backend.h

```diff
@@ -0,0 +1,445 @@
+                                       wLength, timeout);
+    }
+    s32 SubmitTransfer(libusb_transfer* transfer) override {
+        return libusb_submit_transfer(transfer);
+    }
+
+    s32 TryLockEvents() override {
+        return libusb_try_lock_events(g_libusb_context);
+    }
+    void LockEvents() override {
+        return libusb_lock_events(g_libusb_context);
+    }
+    void UnlockEvents() override {
+        return libusb_unlock_events(g_libusb_context);
+    }
+    s32 EventHandlingOk() override {
+        return libusb_event_handling_ok(g_libusb_context);
+    }
+    s32 EventHandlerActive() override {
+        return libusb_event_handler_active(g_libusb_context);
+    }
+    void LockEventWaiters() override {
+        return libusb_lock_event_waiters(g_libusb_context);
+    }
+    void UnlockEventWaiters() override {
+        return libusb_unlock_event_waiters(g_libusb_context);
+    }
+    s32 WaitForEvent(timeval* tv) override {
+        return libusb_wait_for_event(g_libusb_context, tv);
+    }
+
+    s32 HandleEventsTimeout(timeval* tv) override {
+        return libusb_handle_events_timeout(g_libusb_context, tv);
+    }
+    s32 HandleEvents() override {
+        return libusb_handle_events(g_libusb_context);
+    }
+    s32 HandleEventsLocked(timeval* tv) override {
+        return libusb_handle_events_locked(g_libusb_context, tv);
+    }
+
+    s32 CheckConnected(libusb_device_handle* dev) override {
+        // There's no libusb version of this function.
+        // Simulate by querying data.
+
+        int config;
+        return libusb_get_configuration(dev, &config);
+    }
+
+protected:
+    libusb_context* g_libusb_context = nullptr;
+    bool s_has_removed_driver = false;
+};
+
+struct list_head {
+    struct list_head *prev, *next;
+};
+
+#if defined(_WIN32)
+typedef CRITICAL_SECTION usbi_mutex_t;
+#else
+typedef pthread_mutex_t usbi_mutex_t;
+#endif
+
+struct UsbDeviceHandle {
+    usbi_mutex_t lock;
+    unsigned long claimed_interfaces;
+    struct list_head list;
+    struct libusb_device* dev;
+    int auto_detach_kernel_driver;
+};
+
+struct UsbDevice {
+    volatile long refcnt;
+
+    struct libusb_context* ctx;
+    struct libusb_device* parent_dev;
+
+    uint8_t bus_number;
+    uint8_t port_number;
+    uint8_t device_address;
+    enum libusb_speed speed;
+
+    struct list_head list;
+    unsigned long session_data;
+
+    struct libusb_device_descriptor device_descriptor;
+    volatile long attached;
+};
+
+class UsbEmulatedBackend : public UsbRealBackend {
+public:
+    s64 GetDeviceList(libusb_device*** list) override {
+        auto** fake = static_cast<libusb_device**>(calloc(2, sizeof(libusb_device*)));
+        fake[0] = GetDevice(nullptr);
+        fake[1] = nullptr;
+        *list = fake;
+
+        return 1;
+    }
+    void FreeDeviceList(libusb_device** list, s32 unref_devices) override {
+        libusb_free_device_list(list, 0); // to avoid unref fake device
```

Comment:
I believe we should pass unref_devices to not leak the devices in the list (otherwise it only free the list holder)

Reply:
if it makes things simpler, we could use static/global devices for now